File: 1976-number-of-ways-to-arrive-at-destination/1976-number-of-ways-to-arrive-at-destination.java
import java.util.*;

class Solution {
    public static class Edge {
        int src;
        int dest;
        int time;

        public Edge(int src, int dest, int time) {
            this.src = src;
            this.dest = dest;
            this.time = time;
        }
    }

    public int countPaths(int n, int[][] roads) {
        // Build adjacency list using a separate creation() method
        List<Edge>[] graph = creation(n, roads);

        long[] dist = new long[n]; // shortest time to each node
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[0] = 0;

        long[] ways = new long[n]; // number of ways to reach each node
        ways[0] = 1;

        int MOD = 1_000_000_007;

        PriorityQueue<long[]> pq = new PriorityQueue<>(Comparator.comparingLong(a -> a[0]));
        pq.add(new long[]{0, 0}); // {currentTime, node}

        while (!pq.isEmpty()) {
            long[] curr = pq.poll();
            long time = curr[0];
            int node = (int) curr[1];

            if (time > dist[node]) continue; // skip outdated entry

            for (Edge e : graph[node]) {
                int next = e.dest;
                long newTime = time + e.time;

                if (newTime < dist[next]) {
                    dist[next] = newTime;
                    ways[next] = ways[node]; // new shortest path found
                    pq.add(new long[]{newTime, next});
                } else if (newTime == dist[next]) {
                    ways[next] = (ways[next] + ways[node]) % MOD; // additional shortest path
                }
            }
        }

        return (int) ways[n - 1];
    }

    // Separate method to create adjacency list
    private List<Edge>[] creation(int n, int[][] roads) {
        List<Edge>[] graph = new ArrayList[n];
        for (int i = 0; i < n; i++) graph[i] = new ArrayList<>();
        for (int[] road : roads) {
            int u = road[0], v = road[1], t = road[2];
            graph[u].add(new Edge(u, v, t));
            graph[v].add(new Edge(v, u, t)); // bidirectional
        }
        return graph;
    }
}

File: 1976-number-of-ways-to-arrive-at-destination/1976-number-of-ways-to-arrive-at-destination.py
from typing import List
import heapq

class Solution:
    class Edge:
        def __init__(self, src: int, dest: int, time: int):
            self.src = src
            self.dest = dest
            self.time = time

    def countPaths(self, n: int, roads: List[List[int]]) -> int:
        graph = self.creation(n, roads)
        dist = [float('inf')] * n
        dist[0] = 0
        ways = [0] * n
        ways[0] = 1
        MOD = 10**9 + 7

        pq = [(0, 0)]  # (currentTime, node)
        while pq:
            time, node = heapq.heappop(pq)
            if time > dist[node]:
                continue
            for e in graph[node]:
                newTime = time + e.time
                if newTime < dist[e.dest]:
                    dist[e.dest] = newTime
                    ways[e.dest] = ways[node]
                    heapq.heappush(pq, (newTime, e.dest))
                elif newTime == dist[e.dest]:
                    ways[e.dest] = (ways[e.dest] + ways[node]) % MOD

        return ways[n - 1] % MOD

    def creation(self, n: int, roads: List[List[int]]) -> List[List['Solution.Edge']]:
        graph = [[] for _ in range(n)]
        for u, v, t in roads:
            graph[u].append(Solution.Edge(u, v, t))
            graph[v].append(Solution.Edge(v, u, t))  # bidirectional
        return graph
